elks/arch/i86/kernel/irq.c

```diff
@@ -21,75 +21,85 @@
 #include <linuxmt/sched.h>
 #include <linuxmt/timer.h>
 #include <linuxmt/types.h>
+#include <linuxmt/heap.h>
 
 #include <arch/ports.h>
+#include <arch/segment.h>
+#include <arch/irq.h>
 
-/* Enable 80386 Traps */
-/*#define ENABLE_TRAPS */
-
-struct irqaction {
-	void (*handler)();
-	void *dev_id;
-};
-
-static void default_handler(int i, void *regs, void *dev);
-
-static struct irqaction irq_action[] = {
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL},
-#ifdef ENABLE_TRAPS
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL}, {default_handler, NULL}, {default_handler, NULL},
-    {default_handler, NULL},
-#endif
-};
+
+static irq_handler irq_action [16];
 
 /*
  *	Called by the assembler hooks
  */
 
 void do_IRQ(int i,void *regs)
 {
-    register struct irqaction *irq = irq_action + i;
-
-    irq->handler(i, regs, irq->dev_id);
+    register irq_handler ih = irq_action [i];
+    if (!ih) {
+        printk("Unexpected interrupt: %u\n", i);
+        return;
+    }
+	(*ih)(i, regs);
 }
 
-static void default_handler(int i, void *regs, void *dev)
-{
-#ifdef ENABLE_TRAPS
-    if (i > 15)
-	printk("Unexpected trap: %u\n", i-16);
-    else
-#endif
-	printk("Unexpected interrupt: %u\n", i);
-}
 
-int request_irq(int irq, void (*handler)(int,struct pt_regs *,void *), void *dev_id)
+// TODO: simplify the whole by replacing IRQ by INT number
+// Would also allow to handle any of the 0..255 interrupts
+// including the 0..7 processor exceptions & traps
```

Comment:
What is plan here? Will that be part of this PR?

I haven't checked to see whether upper limit on IRQ/INT number is being checked, should probably be to avoid array out of bounds.

Reply:
No plan here, just some ideas... I realized through this PR that INT management and IRQ management are not exactly the same things. For example, installing the INT 80h for system calls, or the CPU traps. So I am still thinking about it, and placing reminders in the code, without changing further.